pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,493 @@
+	pingInterval := time.Duration(nlm.Spec.PingIntervalSeconds) * time.Second
+
+	// Update the latency config
+	m.latencyConfig = &LatencyConfig{
+		Enable:   true,
+		Interval: pingInterval,
+	}
+
+	// Notify the latency config changed
+	m.latencyConfigChanged <- struct{}{}
+
+	return nil
+}
+
+// onNodeLatencyMonitorDelete is the event handler for deleting NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) onNodeLatencyMonitorDelete(obj interface{}) {
+	// Update the latency config
+	m.latencyConfig = &LatencyConfig{Enable: false}
+
+	// Notify the latency config changed
+	m.latencyConfigChanged <- struct{}{}
+}
+
+func (m *NodeLatencyMonitor) sendPing(socket net.PacketConn, addr net.IP) error {
+	var requestType icmp.Type
+
+	// Resolve the IP address
+	ip := &net.IPAddr{IP: addr}
+
+	// Create a new ICMP packet
+	if addr.To4() == nil {
+		requestType = ipv6.ICMPTypeEchoRequest
+	} else {
+		requestType = ipv4.ICMPTypeEcho
+	}
+
+	timeStart := time.Now()
+	seqID := getICMPSeq()
+	body := &icmp.Echo{
+		ID:   os.Getpid() & 0xffff,                       // Use the current process ID as the ICMP ID
+		Seq:  int(seqID),                                 // Use the current sequence number as the ICMP sequence number
+		Data: []byte(timeStart.Format(time.RFC3339Nano)), // Store the current time in the ICMP data
+	}
+	msg := icmp.Message{
+		Type: requestType,
+		Code: 0,
+		Body: body,
+	}
+	klog.InfoS("Send ICMP message", "IP", ip, "SeqID", seqID, "body", body)
+
+	// Serialize the ICMP message
+	msgBytes, err := msg.Marshal(nil)
+	if err != nil {
+		return err
+	}
+
+	// Send the ICMP message
+	_, err = socket.WriteTo(msgBytes, ip)
+	if err != nil {
+		return err
+	}
+
+	// Store current send info
+	// Read the ICMP message
+	oldEntry := m.latencyStore.GetNodeIPLatencyEntryByKey(addr.String())
+	if oldEntry != nil {
+		// Update the latency store
+		oldEntry.LastSendTime = timeStart
+		oldEntry.SeqID = seqID
+		m.latencyStore.UpdateNodeIPLatencyEntryByKey(addr.String(), oldEntry)
+		return nil
+	} else {
+		// New entry
+		m.latencyStore.UpdateNodeIPLatencyEntryByKey(addr.String(), &NodeIPLatencyEntry{
+			SeqID:           seqID,
+			LastSendTime:    timeStart,
+			LastRecvTime:    time.Time{},
+			LastMeasuredRTT: 0,
+		})
+	}
+
+	return nil
+}
+
+func (m *NodeLatencyMonitor) recvPing(socket net.PacketConn, isIPv4 bool, stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			return
+		default:
+			readBuffer := make([]byte, 1500)
+			n, peer, err := socket.ReadFrom(readBuffer)
+			if err != nil {
+				if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
+					klog.ErrorS(err, "Timeout reading ICMP message")
+					continue
+				}
+				klog.ErrorS(err, "Failed to read ICMP message")
+			}
+
+			destIP := peer.String()
+			// Get the node name by destIP
+			entry := m.latencyStore.GetNodeIPLatencyEntryByKey(destIP)
+			if entry == nil {
+				klog.Warning("Failed to get node entry by destIP", "destIP", destIP)
```

Comment:
we don't use `klog.Warning` any more since switching to structured logging. You should either use `klog.InfoS` or `klog.ErrorS` based on what's more appropriate.

Reply:
Ok, I have updated all `klog.Warning` to `klog.V(4).InfoS`.